src/txHashBuilder.c

```diff
@@ -288,20 +292,358 @@ void txHashBuilder_addCertificate_delegation(
+			{
+				BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, p->marginNumerator);
+			}
+			{
+				BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, p->marginDenominator);
+			}
+		}
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, SIZEOF(p->rewardAccount));
+			BUILDER_APPEND_DATA(p->rewardAccount, SIZEOF(p->rewardAccount));
+		}
+	}
+
+	builder->state = TX_HASH_BUILDER_IN_CERTIFICATES_POOL;
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_enterOwners(tx_hash_builder_t* builder)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_enterOwners: %d", builder->state);
+
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL);
+
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, builder->poolCertificateData.remainingOwners);
+	}
+
+	builder->state = TX_HASH_BUILDER_IN_CERTIFICATES_POOL_OWNERS;
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_addOwner(
+        tx_hash_builder_t* builder,
+        const uint8_t* stakingKeyHash, size_t stakingKeyHashSize
+)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_addOwner: %d", builder->state);
+
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_OWNERS);
+	ASSERT(builder->poolCertificateData.remainingOwners > 0);
+	builder->poolCertificateData.remainingOwners--;
+
+	// Bytes[poolKeyHash]
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, stakingKeyHashSize);
+		BUILDER_APPEND_DATA(stakingKeyHash, stakingKeyHashSize);
+	}
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_enterRelays(tx_hash_builder_t* builder)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_enterRelays: %d", builder->state);
+
+	// enter empty owners if none were received
+	if (builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL)
+		txHashBuilder_addPoolRegistrationCertificate_enterOwners(builder);
+
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_OWNERS);
+	ASSERT(builder->poolCertificateData.remainingOwners == 0);
+
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, builder->poolCertificateData.remainingRelays);
+	}
+
+	builder->state = TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS;
+}
+
+static void addRelay_updateState(tx_hash_builder_t* builder)
+{
+	switch (builder->state) {
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL:
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL_OWNERS:
+		txHashBuilder_addPoolRegistrationCertificate_enterRelays(builder);
+		break;
+
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS:
+		break; // we want to be here
+
+	default:
+		ASSERT(false);
+	}
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS);
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_addRelay0(
+        tx_hash_builder_t* builder,
+        const uint16_t* port,
+        const uint8_t* ipv4, size_t ipv4Size,
+        const uint8_t* ipv6, size_t ipv6Size
+)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_addRelay0: %d", builder->state);
+
+	addRelay_updateState(builder);
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS);
```

Comment:
Can relay0, relay1, ... be entered in *any* order?

Reply:
I am not aware of any restriction on relay ordering. We are adding them to the tx body as they come (each in its own APDU). We are not doing much with the relays (not displayed to the user and not fully validated); the sw wallet is responsible for giving us something valid.